dkg/frost.go

```diff
@@ -0,0 +1,275 @@
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package dkg
+
+import (
+	"context"
+
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/coinbase/kryptology/pkg/dkg/frost"
+	"github.com/coinbase/kryptology/pkg/sharing"
+	"github.com/coinbase/kryptology/pkg/signatures/bls/bls_sig"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+// msgKey identifies the source and target nodes and validator index the message belongs to.
+type msgKey struct {
+	// ValIdx identifies the distributed validator (Ith parallel participant) the message belongs to.
+	ValIdx uint32
+
+	// SourceID identifies the source node/participant ID of the message.
+	// It is 1-indexed and equivalent to `cluster.NodeIdx.ShareIdx`.
+	SourceID uint32
+
+	// TargetID identifies the target node/participant ID of the message.
+	// It is 1-indexed and equivalent to `cluster.NodeIdx.ShareIdx`.
+	// The zero value indicates outgoing broadcast messages.
+	TargetID uint32
+}
+
+// fTransport abstracts the transport of frost DKG messages.
+type fTransport interface {
+	// Round1 returns results of all round 1 communication; the received round 1 broadcasts from all other nodes
+	// and the round 1 P2P sends to this node.
+	Round1(context.Context, map[msgKey]frost.Round1Bcast, map[msgKey]sharing.ShamirShare) (
+		map[msgKey]frost.Round1Bcast, map[msgKey]sharing.ShamirShare, error)
+
+	// Round2 returns results of all round 2 communication; the received round 2 broadcasts from all other nodes.
+	Round2(context.Context, map[msgKey]frost.Round2Bcast) (map[msgKey]frost.Round2Bcast, error)
+}
+
+// runFrostParallel runs numValidators Frost DKG processes in parallel (sharing transport rounds)
+// and returns a list of shares (one for each distributed validator).
+//nolint:deadcode // Will be tested and integrated in subsequent PRs.
+func runFrostParallel(ctx context.Context, tp fTransport, numValidators, numNodes, threshold, shareIdx uint32, dgkCtx string) ([]share, error) {
+	validators, err := newFrostParticipants(numValidators, numNodes, threshold, shareIdx, dgkCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	castR1, p2pR1, err := round1(validators)
+	if err != nil {
+		return nil, err
+	}
+
+	castR1Result, p2pR1Result, err := tp.Round1(ctx, castR1, p2pR1)
+	if err != nil {
+		return nil, errors.Wrap(err, "transport round 1")
+	}
+
+	castR2, err := round2(validators, castR1Result, p2pR1Result)
+	if err != nil {
+		return nil, err
+	}
+
+	castR2Result, err := tp.Round2(ctx, castR2)
+	if err != nil {
+		return nil, errors.Wrap(err, "transport round 2")
+	}
+
+	return makeShares(validators, castR2Result)
+}
+
+// newFrostParticipant returns multiple frost dkg participants (one for each parallel validator).
+func newFrostParticipants(numValidators, numNodes, threshold, shareIdx uint32, dgkCtx string) (map[uint32]*frost.DkgParticipant, error) {
```

Comment:
@corverroos in my opinion it is more readable if all the parameters have the type explicit, but I have seen both styles

Reply:
I'm trying to keep the function signature on a single line.